src/MongoDB.Driver/Linq/Linq3Implementation/Translators/TranslationContext.cs

```diff
@@ -14,55 +14,97 @@
+using MongoDB.Bson.Serialization;
 using MongoDB.Driver.Core.Misc;
+using MongoDB.Driver.Linq.Linq3Implementation.Ast.Expressions;
 using MongoDB.Driver.Linq.Linq3Implementation.Misc;
 
 namespace MongoDB.Driver.Linq.Linq3Implementation.Translators
 {
     internal class TranslationContext
     {
         // private fields
+        private readonly NameGenerator _nameGenerator;
         private readonly SymbolTable _symbolTable;
 
         // constructors
         public TranslationContext()
-            : this(new SymbolTable())
+            : this(new SymbolTable(), new NameGenerator())
         {
         }
 
-        public TranslationContext(SymbolTable symbolTable)
+        public TranslationContext(SymbolTable symbolTable, NameGenerator nameGenerator)
         {
             _symbolTable = Ensure.IsNotNull(symbolTable, nameof(symbolTable));
+            _nameGenerator = Ensure.IsNotNull(nameGenerator, nameof(nameGenerator));
         }
 
         // public properties
+        public NameGenerator NameGenerator => _nameGenerator;
         public SymbolTable SymbolTable => _symbolTable;
 
         // public methods
+        public Symbol CreateSymbol(ParameterExpression parameter, IBsonSerializer serializer, bool isCurrent = false)
+        {
+            var parameterName = _nameGenerator.GetParameterName(parameter);
+            return CreateSymbol(parameter, name: parameterName, serializer, isCurrent);
+        }
+
+        public Symbol CreateSymbol(ParameterExpression parameter, string name, IBsonSerializer serializer, bool isCurrent = false)
+        {
+            var varName = _nameGenerator.GetVarName(name);
+            return CreateSymbol(parameter, name, varName, serializer, isCurrent);
+        }
+
+        public Symbol CreateSymbol(ParameterExpression parameter, string name, string varName, IBsonSerializer serializer, bool isCurrent = false)
+        {
+            var varAst = AstExpression.Var(varName, isCurrent);
+            return CreateSymbol(parameter, name, varAst, serializer, isCurrent);
+        }
+
+        public Symbol CreateSymbol(ParameterExpression parameter, AstExpression ast, IBsonSerializer serializer, bool isCurrent = false)
+        {
+            var parameterName = _nameGenerator.GetParameterName(parameter);
+            return CreateSymbol(parameter, name: parameterName, ast, serializer, isCurrent);
+        }
+
+        public Symbol CreateSymbol(ParameterExpression parameter, string name, AstExpression ast, IBsonSerializer serializer, bool isCurrent = false)
+        {
+            return new Symbol(parameter, name, ast, serializer, isCurrent);
+        }
+
+        public Symbol CreateSymbolWithVarName(ParameterExpression parameter, string varName, IBsonSerializer serializer, bool isCurrent = false)
+        {
+            var parameterName = _nameGenerator.GetParameterName(parameter);
+            return CreateSymbol(parameter, name: parameterName, varName, serializer, isCurrent);
+        }
```

Comment:
The general flow is always `CreateSymbol` (one or more times) followed by `WithSymbol(s)`. I couldn't find an example of creating a `Symbol` and then not associating it with the `TranslationContext`. Consider combining `CreateSymbol` and `WithSymbol` into a single method that creates the new symbol and returns a new context containing it.

Reply:
Good enough for me. Thanks for considering it.